### DISCUSSION TITLE
Pipelines, parameters, flags - could we have less?

### DISCUSSION THREAD

**Discussion by ArmoredPony** [nushell/nushell#14163] [General]
Can we theoretically use fewer methods of feeding data to commands? We have a variety of ways to do this: we can use parameters, we can make them optional or use default values, we can use the "rest" syntax, we have flags that are also optional, and in addition, we have pipes that can be used to pipe not only individual parameters. but lists or records, mimicking some of the ways I've listed. Pipes also enable ad hoc polymorphism, making them even more flexible. But some commands can't have their input piped into them, and instead the special `$in` variable must be used, which is for example how we interact with `rm`.

I'm not talking about making any breaking changes here, I'm just asking you to think: if you were a language designer, what - if anything - would you change in this system to reduce ways of invoking a command?

**Comment by fdncred** [comment]
I've always thought that (nearly) every command should allow piped in input and parameterized input to be as flexible as possible.

One goal that we've always had is to make nushell as composable as possible. So, sometimes we take parameters and make new commands from them, then remove those parameters. An example of this is that for a while we had `each` and `for` have a `-i` parameter that would allow you to access the index and the value. We removed those in favor of the `enumerate` command. So, generally speaking, we've added more commands in order to remove parameters. I think there are probably still some commands we could do that with.

**Comment by ghost** [comment]
the problem here is that nushell is not a completely new language.  
shells are a established thing and in order to keep compatability (with `grep`, `ffmpeg`, and other external tools) pipes and rest-parameters are required.  
having flags as their own thing also makes sense and positional (including optional) parameters make things easier in my experience.  
so i would end up with the same 3 systems there.

rm taking `$in` it feels like a backward-compatability thing since that's how it would work in bash, etc (if it had the `$in` variable).  
it might make sense to just allow a bunch of things, but im not sure when it would be to much magic and hurt the language.

> if you were a language designer

if i could throw everything overboard and everything would magically adapt i still would choose to keep pipelines since i like the workflow.  
i might think about replacing parameters with python-style named parameters (`ls long=true dir=.`), but i like the readability of `--` in this context (a language without brackets, etc).

and if we take a look at non-shell scripting languages (python, lua, ecma, etc) we often end up with a very similar design: optional parameters often can be used like parameters (although without the `--` prefix) and `animal.get_age().sqrt()` is basically a pipeline.
